### ISSUE TITLE
Optimize allowedOriginsToDestinations[combinedId]

### ISSUE THREAD

**Issue by jsamuel** [RequestPolicy/requestpolicy#236]
```
imported trac ticket
created: 2011-06-24 13:23:18
reporter: eibwen
```

Origin to Destination policies are implemented as an exists test:
{{{
  isAllowedOriginToDestination : function isAllowedOriginToDestination(
      originIdentifier, destIdentifier) {
    var combinedId = this._getCombinedOriginToDestinationIdentifier(
        originIdentifier, destIdentifier);
    return this._allowedOriginsToDestinations[combinedId] ? true : false;
  },
}}}
or effectively:
{{{
  allowedOriginsToDestinations = {};
  return allowedOriginsToDestinations['Origin' + '|' + 'Destination'] ? true : false;
}}}
This uses string concatenation and an object property lookup.
As an alternative, what about something like:
{{{
  allowedOriginsToDestinations = {};
  return allowedOriginsToDestinations['Origin']['Destination'] ? true : false;
}}}
This uses two object property lookups and no string concatenation.
Benchmarking the two for comparison using http://benchmarkjs.com/:
{{{
  // Static setup
  var o = 'http://example.com';
  var d = 'http://example.org';
  var od = o + '|' + d;
  var o1 = { 'http://example.com|http://example.org': 1 };
  var o2 = { 'http://example.com': { 'http://example.org': 1 }};
  var r;

  // Benchmark [od](asserting o+'|'+d was freely computed)
  r = o1[od] ? true : false;

  // Benchmark [o|d]
  r = o1[o + '|' + d] ? true : false;

  // Benchmark [o][d]
  r = o2[o][d] ? true : false;
}}}
Results (1-to-1):
{{{
  [od]   x 2,316,449 ops/sec ± 0.34% (65 runs sampled)
  [o|d]  x   568,869 ops/sec ±28.24% (50 runs sampled)
  [o][d] x 3,866,819 ops/sec ± 0.20% (68 runs sampled)
  Fastest is [o][d]
}}}
And benchmarking with a larger policy set:
{{{
  // Changing only o1 and o2 (all other lines as above)
  var o1 = {
    'http://example.com|http://ads.example.org': 1,
    'http://example.com|http://example.org': 1,
    'http://example.com|http://img.example.org': 1,
    'http://example.com|http://js.example.org': 1,
    'http://example.com|http://metrics.example.org': 1,
    'http://example.com|http://static.example.org': 1,
    'http://example.com|http://www.example.org': 1
  };
  var o2 = { 'http://example.com': {
    'http://ads.example.org': 1,
    'http://example.org': 1,
    'http://img.example.org': 1,
    'http://js.example.org': 1,
    'http://metrics.example.org': 1,
    'http://static.example.org': 1,
    'http://www.example.org': 1
  }};
}}}
Results (1-to-many):
{{{
  [od]   x 2,292,615 ops/sec ± 0.27% (67 runs sampled)
  [o|d]  x   583,975 ops/sec ±25.89% (46 runs sampled)
  [o][d] x 3,714,361 ops/sec ± 0.30% (69 runs sampled)
  Fastest is [o][d]
}}}
And results with a real world list (many-to-many):
{{{
  [od]   x 2,228,758 ops/sec ± 0.42% (44 runs sampled)
  [o|d]  x   558,977 ops/sec ±29.31% (43 runs sampled)
  [o][d] x 3,723,941 ops/sec ± 0.29% (72 runs sampled)
  Fastest is [o][d]
}}}
And results with a real world list (many-to-many) with return instead of assignment:
{{{
  [od]   x 2,206,056 ops/sec ± 0.34% (68 runs sampled)
  [o|d]  x   630,366 ops/sec ±19.77% (56 runs sampled)
  [o][d] x 3,712,734 ops/sec ± 0.37% (73 runs sampled)
  Fastest is [o][d]
}}}
Conclusions:
- string concatenation is much slower than object property lookups
- benchmark is largely independent of the size of the object
- [od] only had one property lookup to do, but was slower than [o][d] with two property lookups.  Observing that [od]'s property names were longer, and that for one o there could be many d, [o][d] is much more efficient.

An alternative data structure may be faster still; however, the current [o|d] implementation is _slow_!  For now, recommending implementation of [o][d], unless someone can propose something faster (and still readily maintained/understood)...

**Comment by jsamuel** [comment]
```
imported trac comment
created: 2011-06-24 14:10:30
author: eibwen
```

Recognizing that o[p1][p2] throws {{{TypeError}}} if o[p1] is undefined, revising [o][d]:
{{{
  // Benchmark [od]
  return !!o1[od];

  // Benchmark [o|d]
  return !!o1[o + '|' + d];

  // Benchmark [o][d]-if-!!
  if (!o2[o]) return false;
  return !!o2[o][d];

  // Benchmark [o][d]-!!&&!!
  return !!o2[o] && !!o2[o][d];
}}}
Benchmarks with the many-to-many data set:

Results (o and d exist):
{{{
  [od]          x 2,340,441 ops/sec ± 0.33% (62 runs sampled)
  [o|d]         x   639,562 ops/sec ±20.40% (56 runs sampled)
  [o][d]-if-!!  x 2,689,390 ops/sec ± 0.44% (70 runs sampled)
  [o][d]-!!&&!! x 2,680,183 ops/sec ± 0.51% (71 runs sampled)
  Fastest is [o][d]-if-!!,[o][d]-!!&&!!
}}}
Results (o does not exist, d exists; o={{{http://xyzzy.com}}}):
{{{
  [od]          x 279,609 ops/sec ±2.58% (62 runs sampled)
  [o|d]         x 185,715 ops/sec ±1.31% (60 runs sampled)
  [o][d]-if-!!  x 367,450 ops/sec ±1.30% (68 runs sampled)
  [o][d]-!!&&!! x 379,445 ops/sec ±1.19% (68 runs sampled)
  Fastest is [o][d]-!!&&!!,[o][d]-if-!!
}}}
Results (o exists, d does not exist; d={{{http://xyzzy.org}}}):
{{{
  [od]          x 295,534 ops/sec ±2.21% (67 runs sampled)
  [o|d]         x 174,924 ops/sec ±1.25% (56 runs sampled)
  [o][d]-if-!!  x 282,183 ops/sec ±1.09% (68 runs sampled)
  [o][d]-!!&&!! x 296,786 ops/sec ±1.03% (66 runs sampled)
  Fastest is [o][d]-!!&&!!,[od],[o][d]-if-!!
}}}
Results (o does not exist, d does not exist; o={{{http://foobar.com}}}; d={{{http://xyzzy.org}}}):
{{{
  [od]          x 270,271 ops/sec ±2.13% (58 runs sampled)
  [o|d]         x 170,731 ops/sec ±1.93% (59 runs sampled)
  [o][d]-if-!!  x 359,062 ops/sec ±1.02% (67 runs sampled)
  [o][d]-!!&&!! x 387,362 ops/sec ±0.62% (68 runs sampled)
  Fastest is [o][d]-!!&&!!
}}}
Conclusions:
- [o][d](and revised implementations) are still faster overall.
- [od] is comparably fast in the d does not exist scenario; however, it is slower in the other scenarios, and o+'|'+d is unlikely to be free, and [o|d] is slowest in all scenarios.

Lastly, note that both the currently implemented ?: and the proposed !! both depend on the truthiness of the value.  For a pure exists test, independent of value, we would need to use {{{in}}} or {{{hasOwnProperty}}}:
{{{
  // Benchmark [od]
  return !!o1[od];

  // Benchmark [o|d]
  return !!o1[o + '|' + d];

  // Benchmark [o][d]-if-!!
  if (!o2[o]) return false;
  return !!o2[o][d];

  // Benchmark [o][d]-!!&&!!
  return !!o2[o] && !!o2[o][d];

  // Benchmark hasOwnProperty
  return o2.hasOwnProperty(o) && o2[o].hasOwnProperty(d);

  // Benchmark in
  return o in o2 && d in o2[o];

  // Benchmark has-in
  return o2.hasOwnProperty(o) && d in o2[o];

  // Benchmark in-has
  return o in o2 && o2[o].hasOwnProperty(d);
}}}
Results (neither o nor d exist; o={{{http://foobar.com}}}; d={{{http://xyzzy.org}}}):
{{{
  [od]           x   219,911 ops/sec ±3.99% (60 runs sampled)
  [o|d]          x   183,428 ops/sec ±2.08% (61 runs sampled)
  [o][d]-if-!!   x   378,146 ops/sec ±5.66% (59 runs sampled)
  [o][d]-!!&&!!  x   247,911 ops/sec ±2.54% (64 runs sampled)
  hasOwnProperty x 7,336,024 ops/sec ±0.26% (69 runs sampled)
  in             x 7,864,188 ops/sec ±0.44% (68 runs sampled)
  has-in         x 7,338,986 ops/sec ±0.29% (70 runs sampled)
  in-has         x 7,861,154 ops/sec ±0.44% (67 runs sampled)
  Fastest is in,in-has
}}}
Recommending implementation of [o][d] data structure with '''in''' testing provided properties are deleted (and falsy values are never stored).  Otherwise, recommending implementation of either '''{{{[o][d]-if-!!}}}''' or '''{{{[o][d]-!!&&!!}}}'''.

**Comment by jsamuel** [comment]
```
imported trac comment
created: 2011-09-05 12:33:00
author: justin
```

Like many other open tickets, this won't be fixed but instead will become invalid once 1.0 is released.
